### ISSUE TITLE
Compare rule "Watchlisted by" against empty ruleset

### ISSUE THREAD

**Issue by TuGeek** [jorenn92/Maintainerr#1249]
**Describe the bug**
I'm trying to create a rule that says "if 'Watchlisted by' is empty [] then true to delete it" but when I compare the empty list with the string "[]" it tells me it's false.

**To Reproduce**
![image](https://github.com/user-attachments/assets/633117ab-fbc8-40c9-95f2-213760b1088a)

**Expected behavior**
Since the list is empty and I compare it with an empty list, it should give "true" in the rule.
![image](https://github.com/user-attachments/assets/106f12a5-91e0-416e-83a6-a84827fc8bf9)

**Device (please complete the following information):**
 - OS: Docker
 - Version 2.1.0

**Comment by ydkmlt84** [comment]
I just had a thought about this and wanted to share. What if you said watchlisted by, not contains, <some random name>. Then I would assume it would return as true.

**Comment by TuGeek** [comment]
> I just had a thought about this and wanted to share. What if you said watchlisted by, not contains, . Then I would assume it would return as true.

Thats the result with different combinations:
(equals)
![image](https://github.com/user-attachments/assets/3713cf54-a8e9-4d7a-b52e-0cdef1038faa)

(not_equals)
![image](https://github.com/user-attachments/assets/13d421bd-1a2d-4abc-94b7-4fe4b55dde7f)

(contains)
![image](https://github.com/user-attachments/assets/641f0b61-b2e9-470f-8952-4fc7f0fa226c)

(not_contains)
![image](https://github.com/user-attachments/assets/db1ba418-e7b8-4115-a118-c61f118b0060)

(contains_partial)
![image](https://github.com/user-attachments/assets/32fbc1f5-ffe8-4bde-87f8-e7e348dc374f)

(not_contains_partial)
![image](https://github.com/user-attachments/assets/b77e81c1-0d0b-4683-a41f-a3291bc87ef7)

**Comment by ydkmlt84** [comment]
Well, in the formatting of my post I used `<>` and when it was actually posted it removed that section. So I will try again. 

 What I was trying to say is, what if you put a random name in there. Equaling no one is the same as not equaling (James, Lisa, Mary, John, Frank). So if you are looking for stuff that is watchlisted by no one, could you say not contains and list everyone?
 So that when it is comparing it will return true if the list is empty because it wouldn't equal/contain their names?

**Comment by TuGeek** [comment]
Now I think I understand you, but in that case I would need to know the list of all possible people to compare it with the list of people who have seen the item. And for that I would need to have something like "All Plex users".

I hope you understood me now... :)

**Comment by ydkmlt84** [comment]
I thought you were just looking for stuff not watchlisted by anyone.

**Comment by TuGeek** [comment]
Yes, that would be what I'm looking for. Can you think of a way to do it?

**Comment by ydkmlt84** [comment]
> Yes, that would be what I'm looking for. Can you think of a way to do it?

Thats what I was trying to say earlier. Equaling no one is the same as not equaling everyone/anyone. So you could say not contains and list everyone. Then it would return true as it wouldn't equal anyone. Instead of trying to equal blank to blank. Change it to not equaling a list of everyone.

**Comment by ydkmlt84** [comment]
> Yes, that would be what I'm looking for. Can you think of a way to do it?

Also are we sure we're even talking about the same thing? Watchlisted and a "list of watches" are different things. Watchlisted is a feature in Plex where you can add something to your watchlist (or your "want to watch list"). 
 Looking for people who watched something is a different thing. Just wanted to clarify.

**Comment by TuGeek** [comment]
It's hard to explain in writing, and we might be a bit lost in translation... I'm not an English speaker and I might be wrong about something.

My idea is to be able to make a rule that allows a certain item to be deleted when it is not included in anyone's Watchlist.

So, my first idea was "if the rule 'Watchlisted by' is empty" it means that no one has it in their watchlist.

I hope I'll finally be able to get my ideas across to you :)

**Comment by ydkmlt84** [comment]
> It's hard to explain in writing, and we might be a bit lost in translation... I'm not an English speaker and I might be wrong about something.
> 
> 
> 
> My idea is to be able to make a rule that allows a certain item to be deleted when it is not included in anyone's Watchlist.
> 
> 
> 
> So, my first idea was "if the rule 'Watchlisted by' is empty" it means that no one has it in their watchlist.
> 
> 
> 
> I hope I'll finally be able to get my ideas across to you :)
> 
> 
> 
> 

I understand completely. We're on the same page there. 
 How you would do it is put everyone's name whose watchlist it could be in, and say not equals. That is the same as equaling empty. 
 What is the exact opposite of being in no one's watchlist?

**Comment by TuGeek** [comment]
I had thought about enabling the comparison of empty arrays and being able to do the reverse, that is, whenever "Watchlisted by" is an empty array "[]" and I compare it with the empty array text "[]" and it is "equal" it should give True. For me it is comparing 2 empty strings, which in essence should be the same. When someone has seen it, the array would be "[John Doe, Jane Doe]" and when comparing with an empty array "[]" with the "equal" rule it would give False

**Comment by ydkmlt84** [comment]
What do you mean when someone has seen it?

**Comment by TuGeek** [comment]
> What do you mean when someone has seen it?

When someone "Watchlisted", sorry... -_-

**Comment by ydkmlt84** [comment]
You should come over to discord and we can DM about it.

**Comment by henryjrobertson** [comment]
I have same issue, I think. Was there was a resolution discussed elsewhere? Ideally, I just want to add a rule so a title won't get deleted if ~anyone~ has it in their watchlist. Trying to match on [] and [] doesn't seem to work when I test with something that has a watchlist attached.

**Comment by ydkmlt84** [comment]
You don't want to delete something if it's on the watchlist of someone, right?

The solution is to put not contains and list all users `["user1","user2","user3", "etc"]`. It's reverse logic. So, equals no one is the same as doesn't equal everyone.

**Comment by ydkmlt84** [comment]
Nice. I am glad you found your answer!

**Comment by bitnimble** [comment]
Heya, the most recent workaround doesn't seem to work anymore (empty list != empty list is returning false for me), so I just wanted to bump this ticket.

I'm a developer and I'm happy to raise a PR to implement this as a feature, if it's reasonable to simply add "Empty list" as a match value, i.e.:

![Image](https://github.com/user-attachments/assets/42a5a0c0-04ed-46be-9132-694e6d39a640)

While doing `!= [user1, user2, user3, ...]` works, it's a bit inelegant and can be very tedious to enumerate every single Plex user (especially if you have a lot of users on your server), not to mention you have to go update the list every time a new Plex user joins the server. I also don't think it's top-of-mind as a solution when somebody wants to simply match against an empty list, as demonstrated by the existence of this Github issue. 

LMK if that seems reasonable and I can go ahead and raise a PR :)

**Comment by ben-propflo** [comment]
> Heya, the most recent workaround doesn't seem to work anymore (empty list != empty list is returning false for me), so I just wanted to bump this ticket.

You shouldn't need to workaround this any more if I'm understanding your scenario correctly. An empty list to empty list comparison should return true when using equals, so you can do that instead.

With regards to your suggestion, something I'd like to implement for lists is Count equals, Count not equals, Count greater & Count smaller.

**Comment by bitnimble** [comment]
Yep, after cloning and reading through the code and finding the `isStringParsableToArray` bit in the comparator, just realised that equals `[]` works just fine. I swear I tried that the first time around when I first started using Maintainerr and it didn't work... maybe I was just crazy. Well, I'm happy at least :) 

re: count - that sounds reasonable and definitely more flexible. Still happy to raise a PR for that if you think that it's a decent first task into the repo :)

**Comment by benscobie** [comment]
> Yep, after cloning and reading through the code and finding the `isStringParsableToArray` bit in the comparator, just realised that equals `[]` works just fine. I swear I tried that the first time around when I first started using Maintainerr and it didn't work... maybe I was just crazy. Well, I'm happy at least :)

Not crazy haha, this was fixed in [v2.7.0](https://github.com/jorenn92/Maintainerr/releases/tag/v2.7.0)

> re: count - that sounds reasonable and definitely more flexible. Still happy to raise a PR for that if you think that it's a decent first task into the repo :)

We'd be very happy to accept that. I just had a **very** brief look at a possible implementation; add a new RuleType (RuleType.TEXT_ARRAY?) with new RulePossibility's (RulePossibility.COUNT_EQUALS etc.) in `rules.constants.ts`. Then update the [list] ones to return that instead of RuleType.TEXT. That can then be used in the UI to show the correct actions. After that it should just be a case of updating `rule.comparator.service.ts` with the new comparisons and some tests. I am @Scoob on the Maintainerr Discord if you have any questions.

**Comment by ydkmlt84** [comment]
Just to put my two cents in here, there is also a new rule in the dev branch that just returns true or false if the item is watchlisted by anyone.

**Comment by benscobie** [comment]
Closing this now as we updated the equals comparison in a previous version to resolve this, and have just released the list count options.
